services/QuillLMS/app/queries/snapshots/period_query.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+module Snapshots
+  class PeriodQuery
+    attr_accessor :admin_id, :timeframe_start, :timeframe_end, :school_ids, :grades
+
+    def self.run(*args)
+      new(*args).run
+    end
+
+    def initialize(admin_id, timeframe_start, timeframe_end, school_ids = nil, grades = nil)
+      @admin_id = admin_id
+      @timeframe_start = timeframe_start
+      @timeframe_end = timeframe_end
+      @school_ids = school_ids
+      @grades = grades
+    end
+
+    def run
+      raise NotImplementedError
+    end
+
+    def run_query
+      QuillBigQuery::Runner.execute(query)
+    end
+
+    def query
```

Comment:
So writing these generic query combiners is always tricky. I'm a bit on the fence about whether it would be better (but more redundant) to have individual queries share only the `where_clause` and explicitly write everything else out (there could be some shared constants). I worry a little bit that overwriting or modifying superclass vars might make the queries harder to parse (but it is less redundant). Maybe worth a discussion.